### PR TITLE
add support for recursive schemas & structs

### PR DESCRIPTION
**Pull request by adrianiacobghiula**

### CODE REVIEW

**Comment by nrwiersma** [comment]
Any particular reason you chose this approach over passing a DecoderContext/EncoderContext through?

**Comment by adrianiacobghiula** [comment]
Just a matter of personal way of coding, I tend to have the functions on a struct rather than having free-floating with large names `createDecoderOfMap()` vs `encoderCreator.ofMap()` because of input parameters, long-names & cohesion

**Comment by nrwiersma** [comment]
Fair. This has however created some inconsistency in the module as a whole. For example `createDefaultDecoder` still exists with the old signature. `(d decoderCreator) newEfaceDecoder` is actually a constructor but has been attached.

**Comment by adrianiacobghiula** [comment]
Pushed improvements. I do have one question: What does `eface` stands for ?

**Comment by nrwiersma** [comment]
Interfaces are split into 2 types, an empty interface (interface{} or any) `eface` and a populated (has methods) interface `iface`.

**Comment by nrwiersma** [comment]
The issue this poses is that it fundamentally changes the codebase far beyond what is required to implement the feature. In fact I see a couple ways to do it without putting everything under a struct. Part of this also subtly changes how this works in ways I don't fully understand the impact of yet. I am not making any argument of the relative benefits of either approach. Both are perfectly valid, and were I to completely rewrite today, perhaps a different approach would be chosen.

As an example, this changes the benchmark of the library for the worse (~224ns to ~230ns), but it would be very difficult to pin down exactly were the degraded performance comes from, small as it may be.

**Comment by adrianiacobghiula** [comment]
So you would not have decoderCreator or encoderCreator but replace the parameter `cfg *frozenConfig` to another object like `DecoderContext` (that contains the `cfg *frozenConfig` & the map ?) to every function `createDecoderOfMap`, `decoderOfType`, `createDecoderOfMarshaler` ?

**Comment by nrwiersma** [comment]
Yes. That seems like it should work.

**Comment by adrianiacobghiula** [comment]
@nrwiersma pushed using `DecoderContext` & `EncoderContext`

File: codec.go [hunk 1]
```diff
@@ -71,48 +71,88 @@ func (c *frozenConfig) DecoderOf(schema Schema, typ reflect2.Type) ValDecoder {
...
 	}
 
 	switch schema.Type() {
 	case String, Bytes, Int, Long, Float, Double, Boolean:
 		return createDecoderOfNative(schema.(*PrimitiveSchema), typ)
-
 	case Record:
-		return createDecoderOfRecord(cfg, schema, typ)
-
+		key := cacheKey{fingerprint: schema.Fingerprint(), rtype: typ.RType()}
```
Comment (nrwiersma) [inline codec.go:135]: I think this should use `CacheFingerprint` which takes defaults into account.

File: codec.go [hunk 2]
```diff
@@ -71,48 +71,88 @@ func (c *frozenConfig) DecoderOf(schema Schema, typ reflect2.Type) ValDecoder {
...
+		defDec := &deferDecoder{}
+		d.decoders[key] = defDec
+		defDec.decoder = createDecoderOfRecord(d, schema.(*RecordSchema), typ)
+		return defDec.decoder
 	case Ref:
-		return decoderOfType(cfg, schema.(*RefSchema).Schema(), typ)
-
+		key := cacheKey{fingerprint: schema.(*RefSchema).Schema().Fingerprint(), rtype: typ.RType()}
```
Comment (nrwiersma) [inline codec.go:141]: Same here. I think it needs to use `CacheFingerprint`.

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+.idea
```
Comment (nrwiersma) [inline .gitignore:1]: This should be moved to your global `.gitignore`. If we support 1 platform, we must support all, and there are very very many.

**Review by nrwiersma — CHANGES_REQUESTED** [review]
Nice addition, thanks. A couple comments.

File: codec.go [hunk 4]
```diff
@@ -158,28 +198,28 @@ func encoderOfType(cfg *frozenConfig, schema Schema, typ reflect2.Type) ValEncod
 	switch schema.Type() {
 	case String, Bytes, Int, Long, Float, Double, Boolean, Null:
 		return createEncoderOfNative(schema, typ)
-
 	case Record:
-		return createEncoderOfRecord(cfg, schema, typ)
-
+		key := cacheKey{fingerprint: schema.CacheFingerprint(), rtype: typ.RType()}
```
Comment (nrwiersma) [inline codec.go:202]: Composite schemas are only a thing for decoding. 
```suggestion
		key := cacheKey{fingerprint: schema.Fingerprint(), rtype: typ.RType()}
```

File: codec.go [hunk 5]
```diff
@@ -158,28 +198,28 @@ func encoderOfType(cfg *frozenConfig, schema Schema, typ reflect2.Type) ValEncod
...
+		defEnc := &deferEncoder{}
+		e.encoders[key] = defEnc
+		defEnc.encoder = createEncoderOfRecord(e, schema.(*RecordSchema), typ)
+		return defEnc.encoder
 	case Ref:
-		return encoderOfType(cfg, schema.(*RefSchema).Schema(), typ)
-
+		key := cacheKey{fingerprint: schema.(*RefSchema).Schema().CacheFingerprint(), rtype: typ.RType()}
```
Comment (nrwiersma) [inline codec.go:208]:
```suggestion
		key := cacheKey{fingerprint: schema.(*RefSchema).Schema().Fingerprint(), rtype: typ.RType()}
```

**Review by nrwiersma — CHANGES_REQUESTED** [review]

**Review by nrwiersma — APPROVED** [review]
LGTM 🎉 Thanks for the contribution.

**Pull request merged**